internal/throttler/remote/throttler.go

```diff
@@ -0,0 +1,178 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"fmt"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/uber-go/atomic"
+
+	"github.com/uber/jaeger-client-go"
+	"github.com/uber/jaeger-client-go/utils"
+)
+
+const (
+	// minimumCredits is the minimum amount of credits necessary to not be throttled.
+	// i.e. if currentCredits > minimumCredits, then the operation will not be throttled.
+	minimumCredits = 1.0
+)
+
+type creditResponse struct {
+	Operation string  `json:"operation"`
+	Credits   float64 `json:"credits"`
+}
+
+type httpCreditManagerProxy struct {
+	hostPort string
+	logger   jaeger.Logger
+}
+
+func newHTTPCreditManagerProxy(hostPort string, logger jaeger.Logger) *httpCreditManagerProxy {
+	return &httpCreditManagerProxy{
+		hostPort: hostPort,
+		logger:   logger,
+	}
+}
+
+func (m *httpCreditManagerProxy) FetchCredits(uuid, serviceName string, operation []string) []creditResponse {
+	params := url.Values{}
+	params.Set("service", serviceName)
+	params.Set("uuid", uuid)
+	for _, op := range operation {
+		params.Add("operation", op)
+	}
+	var resp []creditResponse
+	if err := utils.GetJSON(fmt.Sprintf("http://%s/credits?%s", m.hostPort, params.Encode()), &resp); err != nil {
+		m.logger.Error("Failed to receive credits from agent: " + err.Error())
+	}
+	return resp
+}
+
+// Throttler retrieves credits from agent and uses it to throttle operations.
+type Throttler struct {
+	options
+
+	mux           sync.RWMutex
+	service       string
+	uuid          *atomic.String
+	creditManager *httpCreditManagerProxy
+	credits       map[string]float64 // map of operation->credits
+	close         chan struct{}
+	stopped       sync.WaitGroup
+}
+
+// NewThrottler returns a Throttler that polls agent for credits and uses them to throttle
+// the service.
+func NewThrottler(service string, options ...Option) *Throttler {
+	// TODO add metrics
+	// TODO set a limit on the max number of credits
+	opts := applyOptions(options...)
+	creditManager := newHTTPCreditManagerProxy(opts.hostPort, opts.logger)
+	t := &Throttler{
+		options:       opts,
+		creditManager: creditManager,
+		service:       service,
+		credits:       make(map[string]float64),
+		close:         make(chan struct{}),
+		uuid:          atomic.NewString(""),
+	}
+	t.stopped.Add(1)
+	go t.pollManager()
+	return t
+}
+
+// IsThrottled implements Throttler#IsThrottled.
+func (t *Throttler) IsThrottled(operation string) bool {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	_, ok := t.credits[operation]
+	if !ok {
+		// If it is the first time this operation is being checked, synchronously fetch
+		// the credits.
```

Comment:
> synchronously fetch

can we make this configurable? Only yab needs it synchronously, normal services could just wait for refresh?